tests/test_indexes.py

```diff
@@ -684,3 +688,72 @@ def test_sorted_and_list() -> None:
         ),
         list,
     )
+
+
+def test_index_operators() -> None:
+    # GH 405
+    i1 = pd.Index([1, 2, 3])
+    i2 = pd.Index([4, 5, 6])
+
+    check(assert_type(i1 + i2, pd.Index), pd.Index)
+    check(assert_type(i1 + 10, pd.Index), pd.Index)
+    check(assert_type(10 + i1, pd.Index), pd.Index)
+    check(assert_type(i1 - i2, pd.Index), pd.Index)
+    check(assert_type(i1 - 10, pd.Index), pd.Index)
+    check(assert_type(10 - i1, pd.Index), pd.Index)
+    check(assert_type(i1 * i2, pd.Index), pd.Index)
+    check(assert_type(i1 * 10, pd.Index), pd.Index)
+    check(assert_type(10 * i1, pd.Index), pd.Index)
+    check(assert_type(i1 / i2, pd.Index), pd.Index)
+    check(assert_type(i1 / 10, pd.Index), pd.Index)
+    check(assert_type(10 / i1, pd.Index), pd.Index)
+    check(assert_type(i1 // i2, pd.Index), pd.Index)
+    check(assert_type(i1 // 10, pd.Index), pd.Index)
+    check(assert_type(10 // i1, pd.Index), pd.Index)
+    check(assert_type(i1**i2, pd.Index), pd.Index)
+    check(assert_type(i1**2, pd.Index), pd.Index)
+    check(assert_type(2**i1, pd.Index), pd.Index)
+    check(assert_type(i1 % i2, pd.Index), pd.Index)
+    check(assert_type(i1 % 10, pd.Index), pd.Index)
+    check(assert_type(10 % i1, pd.Index), pd.Index)
+    check(assert_type(divmod(i1, i2), Tuple[pd.Index, pd.Index]), tuple)
+    check(assert_type(divmod(i1, 10), Tuple[pd.Index, pd.Index]), tuple)
+    check(assert_type(divmod(10, i1), Tuple[pd.Index, pd.Index]), tuple)
+
+    if TYPE_CHECKING_INVALID_USAGE:
+        assert_type(
+            i1 & i2,  # type:ignore[operator] # pyright: ignore[reportGeneralTypeIssues]
+            Never,
+        )
+        assert_type(  # type: ignore[assert-type]
+            i1 & 10,  # type:ignore[operator] # pyright: ignore[reportGeneralTypeIssues]
+            Never,
+        )
+        assert_type(  # type: ignore[assert-type]
+            10 & i1,  # type:ignore[operator] # pyright: ignore[reportGeneralTypeIssues]
+            Never,
+        )
+        assert_type(
+            i1 | i2,  # type:ignore[operator] # pyright: ignore[reportGeneralTypeIssues]
+            Never,
+        )
+        assert_type(  # type: ignore[assert-type]
```

Comment:
Some of the checks have an ignore for `assert_type` and some don't. Why do some need an ignore?

Reply:
It has to do with how `mypy` processes `Never` with certain operators and arguments.  If you look carefully, you'll see that the `# ignore[assert-type]` is occurring on any operator that has a numerical operand.  Without the `ignore`, `mypy` says the `assert_type()` is wrong, thinking the result is `Any`.  Probably some kind of `mypy` error, although the interpretation of the `Never` result is ambiguous.